docs/features/non-source-outputs.md

```diff
@@ -0,0 +1,141 @@
+
+### Rationale
+
+There exists a desire for incremental source generators be able to write out a set of accompanying files, alongside the generated sources that they add to the compilation. An example of this is JavaScript interop with .NET code. When exposing a .NET method to JavaScript today it is possible to generate the C# interop source code that enables the call to be made from JavaScript, but the user must still hand author the corresponding JavaScript code. Instead, with this feature the incremental generator could emit both the required C# code alongside a JavaScript file that provides the equivalent JavaScript interop code.
+
+### High Level Design Goals
+
+- Allow Incremental Generators to emit source or binary files to disk alongside the emitted .NET assembly
+- The user can control where the generated artifacts are emitted with a suitable default
+- Be optimal in terms of file writes / copies and ensure we keep the correct incremental build semantics for MSBuild
+- Do not have to be run as part of the IDE experience. Files are only emitted to disk as part of command line compilation
+
+## Implementation
+
+The feature is exposed to incremental generator authors via a new output API:
+
+```csharp
+namespace Microsoft.CodeAnalysis;
+
+public readonly partial struct IncrementalGeneratorInitializationContext
+{
+    public void RegisterArtifactOutput<TSource>(IncrementalValueProvider<TSource> source, Action<ArtifactProductionContext, TSource> action);
+    public void RegisterArtifactOutput<TSource>(IncrementalValuesProvider<TSource> source, Action<ArtifactProductionContext, TSource> action);
+}
+
+public readonly struct ArtifactProductionContext
+{
+    public CancellationToken CancellationToken { get; }
+
+    public void AddFile(string hintName, Action<System.IO.Stream, CancellationToken> callback);
+}
+```
+
+This can be used by a generator author like so:
+
+```csharp
+internal sealed class Generator : IIncrementalGenerator
+{
+    public void Initialize(IncrementalGeneratorInitializationContext context)
+    {
+        var compilationSource = context.CompilationProvider;
+
+        context.RegisterArtifactOutput(compilationSource, (nspc, item) =>
+        {
+            nspc.AddFile("interop.js", (stream, cancellationToken) =>
+            {
+                using var writer = new StreamWriter(stream);
+                writer.WriteLine("//output");
+            });
+        });
+    }
+}
+```
+
+Note how the call to `AddFile` takes a callback that provides a stream. This allows the host of the generator driver to control how the stream is created and what kind of stream is created. For instance in the compiler it could be a `FileStream` that writes directly to disk, but in unit tests it could be a `MemoryStream` that can be inspected as part of the test. It also takes a `hintName`, the same way that source files do, rather than an explicit file name or path. This is explained in more detail below.
```

Comment:
> For instance in the compiler it could be a `FileStream` that writes directly to disk

I don't love this.  Especially as it now opens up gnerator code to concepts like IO exceptions.  I think we should buffer, and write out and own that entire painful space.  Or we should ensure that however we wrap, we deal with the exceptinos. 

Note: testing around exceptions should be done here.

Reply:
This was a trouble spot for Roslyn in our first few releases. While reading the doc I paged back in the code cause my mental model is that this should mostly follow how we write our PE, PDB, etc ... files to disk. 

Indeed we do absolutely use `FileStream` as the implementation. That does mean IO exceptions are a very real problem and it's one the compiler put a lot of work into handling.

Reply:
I would assume that the compiler would handle the exceptions, in the same way we handle exceptions today in the generator driver. We wrap the callbacks, catch, and turn them into generator errors, not raw exceptions.